internal/internal.go

```diff
@@ -279,16 +290,34 @@ func (p *envoyExtAuthzGrpcServer) eval(ctx context.Context, input ast.Value, opt
 			"txn":     result.txnID,
 		}).Infof("Executing policy query.")
 
-		opts = append(opts,
-			rego.Metrics(result.metrics),
-			rego.ParsedQuery(p.cfg.parsedQuery),
-			rego.ParsedInput(input),
-			rego.Compiler(p.manager.GetCompiler()),
-			rego.Store(p.manager.Store),
-			rego.Transaction(txn),
-			rego.Runtime(p.manager.Info))
+		p.preparedQueryMutex.Lock()
```

Comment:
@ashutosh-narkar @patrick-east are we serializing policy evaluation intentionally here? I would have expected the lock to only be held when checking if the prepared query needs to be constructed. Assuming the Envoy and the ext_authz filter can execute concurrent checks, this will become a bottleneck. We could avoid the bottleneck by moving the locking/construction into a helper function.